nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -291,30 +347,197 @@ impl DataStore {
+                        // If we have a "required" sled, it must be chosen.
+
+                        if banned.contains(&required_id) {
+                            return Err(err.bail(SledReservationError::ConflictingAntiAndAffinityConstraints));
+                        }
+                        if !targets.contains(&required_id) {
+                            return Err(err.bail(SledReservationError::RequiredAffinitySledNotValid));
+                        }
+                        *required_id
+                    } else {
+                        // We have no "required" sleds, but might have preferences.
+
+                        targets = targets.difference(&banned).cloned().collect();
+
+                        let mut preferred = affinity_sleds.iter().filter_map(|(policy, id)| {
+                            if *policy == AffinityPolicy::Allow {
+                                Some(*id)
+                            } else {
+                                None
+                            }
+                        }).collect::<HashSet<_>>();
+                        let mut unpreferred = anti_affinity_sleds.iter().filter_map(|(policy, id)| {
+                            if *policy == AffinityPolicy::Allow {
+                                Some(*id)
+                            } else {
+                                None
+                            }
+                        }).collect::<HashSet<_>>();
+
+                        // Only consider "preferred" sleds that are viable targets
+                        preferred = targets.intersection(&preferred).cloned().collect();
+                        // Only consider "unpreferred" sleds that are viable targets
+                        unpreferred = targets.intersection(&unpreferred).cloned().collect();
+
+                        // If a target is both preferred and unpreferred, it is removed
+                        // from both sets.
+                        let both = preferred.intersection(&unpreferred).cloned().collect();
+                        preferred = preferred.difference(&both).cloned().collect();
+                        unpreferred = unpreferred.difference(&both).cloned().collect();
+
+                        if let Some(target) = preferred.iter().next() {
+                            *target
+                        } else {
+                            targets = targets.difference(&unpreferred).cloned().collect();
+                            if let Some(target) = targets.iter().next() {
+                                *target
+                            } else {
+                                if let Some(target) = unpreferred.iter().next() {
+                                    *target
+                                } else {
+                                    return Err(err.bail(SledReservationError::NotFound));
+                                }
+                            }
+                        }
```

Comment:
oh jeez...it's a bummer we can't express this logic with early returns, because we still have to call `SledResource::new_for_vmm` and actually insert the resource record. I kind of wonder whether it's worth factoring out the selection of a sled target into a standalone function that's called by the transaction, so you can use early returns to express this logic with a bit less rightward drift. it might also make it easier to write tests for the sled selection code without having to actually do a wole DB transaction?

Reply:
I'm all in favor of that -- pulled into a free function, reduced rightward drift too.